safekeeper/src/safekeeper.rs

```diff
@@ -851,6 +851,29 @@ where
             return Ok(Some(AcceptorProposerMessage::AppendResponse(resp)));
         }
 
+        // Disallow any non-sequential writes, which can result in gaps or
+        // overwrites. If we need to move the pointer, ProposerElected message
+        // should have truncated WAL first accordingly. Note that the first
+        // condition (WAL rewrite) is quite expected in real world; it happens
+        // when walproposer reconnects to safekeeper and writes some more data
+        // while first connection still gets some packets later. It might be
+        // better to not log this as error! above.
+        let write_lsn = self.wal_store.write_lsn();
+        if write_lsn > msg.h.begin_lsn {
+            bail!(
+                "append request rewrites WAL written before, write_lsn={}, msg lsn={}",
+                write_lsn,
+                msg.h.begin_lsn
+            );
+        }
+        if write_lsn < msg.h.begin_lsn && write_lsn != Lsn(0) {
+            bail!(
```

Comment:
Thinking about this scenario:
1. Client sends some WAL in range LSN 1,2,3 on a tcp socket
2. Client doesn't get acks because the connection dies, but the write actually made it to the safekeeper
3. Client tries to write again from LSN 1, hits this error because it is writing into the past relative to write_lsn

Will the client handle that gracefully?  How will it know what LSN to start writing from?

Reply:
Each compute -> safekeeper connection starts with the voting protocol, even if compute is already elected. During it safekeeper sends its <term, flush_lsn> position (if fact, full history of term switches) to the compute so it could calculate the streaming point. So we might observe this error, but eventually when all old connection(s) go away and one remains it should stabilize.